lib/rules/selector-not-notation/index.js

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * @param {Node} node
+ * @returns {boolean}
+ */
+const isPseudoClass = (node) =>
+	buggyIsPseudoClass(node) &&
+	node.value.toLowerCase() !== ':first-line' &&
+	node.value.toLowerCase() !== ':first-letter';
+
+/**
+ * @param {Node} node
+ * @returns {boolean}
+ */
+const isSimpleSelector = (node) =>
+	isPseudoClass(node) ||
+	isAttribute(node) ||
+	isClassName(node) ||
+	isUniversal(node) ||
+	isIdentifier(node) ||
+	isTag(node);
+
+/**
+ * @param {Node} node
+ * @returns {boolean}
+ */
+const isNot = (node) =>
+	isPseudoClass(node) && node.value !== undefined && node.value.toLowerCase() === ':not';
+
+/**
+ * @param {Selector[]} list
+ * @returns {boolean}
+ */
+const isSimple = (list) => {
+	if (list.length > 1) return false;
+
+	const [first, second] = list[0].nodes; // list is never empty
+
+	if (!first) return true;
+
+	if (second) return false;
+
+	return isSimpleSelector(first) && !isNot(first);
+};
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary, _, context) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: ['simple', 'complex'],
+		});
+
+		if (!validOptions) return;
+
+		root.walkRules((ruleNode) => {
+			if (!isStandardSyntaxRule(ruleNode)) return;
+
+			const selector = ruleNode.selector;
+
+			if (!selector.includes(':not(')) return;
+
+			if (!isStandardSyntaxSelector(selector)) return;
+
+			const fixedSelector = parseSelector(selector, result, ruleNode, (container) => {
+				container.walkPseudos((pseudo) => {
+					if (!isNot(pseudo)) return;
+
+					if (primary === 'complex') {
+						const prev = pseudo.prev();
+						const hasConsecutiveNot = prev && isNot(prev);
+
+						if (!hasConsecutiveNot) return;
+
+						if (context.fix) return fixComplex(/** @type {Pseudo} */ (prev));
+					} else {
+						const selectors = pseudo.nodes;
+
+						if (isSimple(selectors)) return;
+
+						const mustFix =
+							context.fix &&
+							selectors.length > 1 &&
+							(!selectors[1].nodes.length || selectors.every(({ nodes }) => nodes.length === 1));
+
+						if (mustFix) return fixSimple(pseudo);
+					}
+
+					report({
+						message: messages.expected(primary),
+						node: ruleNode,
+						index: pseudo.sourceIndex,
+						result,
+						ruleName,
+					});
+				});
+			});
+
+			if (context.fix && fixedSelector) {
+				ruleNode.selector = fixedSelector;
+			}
+		});
+	};
+};
+
+/**
+ * @param {Pseudo} not
+ */
+function fixSimple(not) {
+	const simpleSelectors = not.nodes
+		.filter(({ nodes }) => isSimpleSelector(nodes[0]))
+		.map((s) => {
+			s.nodes[0].rawSpaceBefore = '';
+			s.nodes[0].rawSpaceAfter = '';
+
+			return s;
+		});
+
+	not.empty();
+	// @ts-ignore
+	not.nodes.push(simpleSelectors.shift());
+
+	for (const s of simpleSelectors) {
+		const last = /** @type {Container} */ (not.parent).last;
+
+		/** @type {Container} */ (not.parent).insertAfter(
```

Comment:
[suggest] We can remove type-cast via `assert()` like this:

```suggestion
	assert(not.parent);

	for (const s of simpleSelectors) {
		const last = not.parent.last;

		not.parent.insertAfter(
```